File: NeoRoleplaybackend-main/app/schemas/session.py
from pydantic import BaseModel
from typing import Optional, Any
from datetime import datetime

from app.models.enums import SessionType, SessionStatus, Framework, TimelineEventType


class ScenarioConfigInput(BaseModel):
    industry: str
    roleplayType: str
    personaContext: str
    displayName: str = "Custom Persona"
    displayTitle: str = ""
    displayEmoji: str = "🧑‍💼"
    difficulty: str = "Medium"
    suggestedQuestions: list[str] = []
    objections: list[str] = []
    aiCanEnd: bool = True
    endCondition: str = ""
    timeLimitMins: Optional[int] = None
    language: Optional[str] = "English"


class ScenarioConfigResponse(BaseModel):
    industry: str
    roleplayType: str
    personaContext: str
    displayName: str
    displayTitle: str
    displayEmoji: str
    difficulty: str
    suggestedQuestions: list[str] = []
    objections: list[str] = []
    aiCanEnd: bool = True
    endCondition: str = ""
    timeLimitMins: Optional[int] = None
    language: Optional[str] = "English"


class CreateSessionRequest(BaseModel):
    personaId: Optional[str] = None
    scenarioConfig: Optional[ScenarioConfigInput] = None
    type: SessionType
    framework: Framework


class EndSessionRequest(BaseModel):
    transcript: Optional[list[Any]] = None
    durationSeconds: int
    recordingUrl: Optional[str] = None
    skipAnalysis: bool = False


class AddMessageRequest(BaseModel):
    role: str
    content: str
    timestampMs: int
    audioUrl: Optional[str] = None


class AIRespondRequest(BaseModel):
    userMessage: str


class PersonaMinimal(BaseModel):
    id: str
    name: str
    title: str
    emoji: str
    difficulty: str

    model_config = {"from_attributes": True}


class UserMinimal(BaseModel):
    id: str
    firstName: str
    lastName: str
    avatarUrl: Optional[str] = None

    model_config = {"from_attributes": True}


class MessageResponse(BaseModel):
    id: str
    sessionId: str
    role: str
    content: str
    audioUrl: Optional[str] = None
    timestampMs: int
    createdAt: datetime

    model_config = {"from_attributes": True}


class FrameworkScoreResponse(BaseModel):
    id: str
    component: str
    score: float
    feedback: str
    evidence: list[str] = []

    model_config = {"from_attributes": True}


class TimelineEventResponse(BaseModel):
    id: str
    type: TimelineEventType
    timestampMs: int
    title: str
    description: str
    suggestion: Optional[str] = None
    transcriptRef: Optional[str] = None
    betterResponse: Optional[str] = None

    model_config = {"from_attributes": True}


class SessionResponse(BaseModel):
    id: str
    type: SessionType
    status: SessionStatus
    framework: Framework
    totalScore: Optional[float] = None
    durationSeconds: Optional[int] = None
    recordingUrl: Optional[str] = None
    startedAt: Optional[datetime] = None
    endedAt: Optional[datetime] = None
    createdAt: datetime
    user: Optional[UserMinimal] = None
    persona: Optional[PersonaMinimal] = None
    scenarioConfig: Optional[ScenarioConfigResponse] = None
    frameworkScores: list[FrameworkScoreResponse] = []
    timelineEvents: list[TimelineEventResponse] = []
    messages: list[MessageResponse] = []

    model_config = {"from_attributes": True}


class PaginationResponse(BaseModel):
    page: int
    limit: int
    total: int
    pages: int


class SessionListResponse(BaseModel):
    sessions: list[SessionResponse]
    pagination: PaginationResponse


class AIRespondResponse(BaseModel):
    response: str

File: backend/src/models/Session.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IMessage {
  _id?: mongoose.Types.ObjectId;
  role: 'user' | 'assistant';
  content: string;
  timestampMs: number;
  audioUrl?: string;
}

export interface IFrameworkScore {
  _id?: mongoose.Types.ObjectId;
  component: string;
  score: number;
  feedback: string;
  evidence: string[];
}

export interface ITimelineEvent {
  _id?: mongoose.Types.ObjectId;
  type: 'ISSUE' | 'GOOD' | 'WARNING' | 'NEUTRAL';
  timestampMs: number;
  title: string;
  description: string;
  suggestion?: string;
  transcriptRef?: string;
  betterResponse?: string;
}

export interface IScenarioConfig {
  industry: string;
  roleplayType: string;
  personaContext: string;
  displayName: string;
  displayTitle: string;
  displayEmoji: string;
  difficulty: string;
  suggestedQuestions: string[];
  objections?: string[];
  aiCanEnd?: boolean;
  endCondition?: string;
  timeLimitMins?: number | null;
  avatarId?: string;
  elevenlabsVoiceId?: string;
  language?: string;
}

export interface ISession extends Document {
  type: 'PHONE_CALL' | 'ONLINE_MEETING';
  status: 'PENDING' | 'IN_PROGRESS' | 'COMPLETED' | 'FAILED';
  framework: string;
  totalScore?: number;
  durationSeconds?: number;
  recordingUrl?: string;
  aiFeedback?: string;
  analysisSkipped?: boolean;
  convaiConversationId?: string;
  startedAt?: Date;
  endedAt?: Date;
  userId: mongoose.Types.ObjectId;
  personaId?: mongoose.Types.ObjectId;
  companyId: mongoose.Types.ObjectId;
  scenarioConfig?: IScenarioConfig;
  messages: IMessage[];
  frameworkScores: IFrameworkScore[];
  timelineEvents: ITimelineEvent[];
  createdAt: Date;
  updatedAt: Date;
}

const MessageSchema = new Schema<IMessage>({
  role: { type: String, enum: ['user', 'assistant'], required: true },
  content: { type: String, required: true },
  timestampMs: { type: Number, required: true },
  audioUrl: String,
});

const FrameworkScoreSchema = new Schema<IFrameworkScore>({
  component: String,
  score: Number,
  feedback: String,
  evidence: [String],
});

const TimelineEventSchema = new Schema<ITimelineEvent>({
  type: { type: String, enum: ['ISSUE', 'GOOD', 'WARNING', 'NEUTRAL'] },
  timestampMs: Number,
  title: String,
  description: String,
  suggestion: String,
  transcriptRef: String,
  betterResponse: String,
});

const ScenarioConfigSchema = new Schema<IScenarioConfig>({
  industry: String,
  roleplayType: String,
  personaContext: String,
  displayName: String,
  displayTitle: String,
  displayEmoji: String,
  difficulty: String,
  suggestedQuestions: [String],
  objections: [String],
  aiCanEnd: Boolean,
  endCondition: String,
  timeLimitMins: Number,
  avatarId: String,
  elevenlabsVoiceId: String,
  language: String,
});

const SessionSchema = new Schema<ISession>(
  {
    type: { type: String, enum: ['PHONE_CALL', 'ONLINE_MEETING'], required: true },
    status: { type: String, enum: ['PENDING', 'IN_PROGRESS', 'COMPLETED', 'FAILED'], default: 'PENDING' },
    framework: { type: String, required: true },
    totalScore: Number,
    durationSeconds: Number,
    recordingUrl: String,
    aiFeedback: String,
    analysisSkipped: Boolean,
    convaiConversationId: String,
    startedAt: Date,
    endedAt: Date,
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    personaId: { type: Schema.Types.ObjectId, ref: 'Persona' },
    companyId: { type: Schema.Types.ObjectId, ref: 'Company', required: true },
    scenarioConfig: ScenarioConfigSchema,
    messages: [MessageSchema],
    frameworkScores: [FrameworkScoreSchema],
    timelineEvents: [TimelineEventSchema],
  },
  { timestamps: true }
);

SessionSchema.index({ userId: 1, createdAt: -1 });
SessionSchema.index({ companyId: 1, createdAt: -1 });

export const Session = mongoose.model<ISession>('Session', SessionSchema);
